cpp/src/arrow/filesystem/path_tree_test.cc

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/filesystem/path_tree.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "arrow/filesystem/path_util.h"
+#include "arrow/filesystem/test_util.h"
+#include "arrow/testing/gtest_util.h"
+
+namespace arrow {
+namespace fs {
+
+static std::shared_ptr<PathTree> PT(FileStats stats) {
+  return std::make_shared<PathTree>(std::move(stats));
+}
+
+static std::shared_ptr<PathTree> PT(FileStats stats,
+                                    std::vector<std::shared_ptr<PathTree>> subtrees) {
+  return std::make_shared<PathTree>(std::move(stats), std::move(subtrees));
+}
+
+void AssertMakePathTree(std::vector<FileStats> stats,
+                        std::vector<std::shared_ptr<PathTree>> expected) {
+  std::vector<std::shared_ptr<PathTree>> actual;
+
+  ASSERT_OK(PathTree::Make(stats, &actual));
+  EXPECT_THAT(actual, testing::ContainerEq(expected));
+}
+
+TEST(TestPathTree, Basic) {
+  AssertMakePathTree({}, {});
+
+  AssertMakePathTree({File("aa")}, {PT(File("aa"))});
+  AssertMakePathTree({Dir("AA")}, {PT(Dir("AA"))});
+  AssertMakePathTree({Dir("AA"), File("AA/aa")}, {PT(Dir("AA"), {PT(File("AA/aa"))})});
+
+  // Missing parent can still find ancestor.
+  AssertMakePathTree({Dir("AA"), File("AA/BB/bb")},
+                     {PT(Dir("AA"), {PT(File("AA/BB/bb"))})});
```

Comment:
This is weird. The intermediate directory (`AA/BB`) isn't even present in the tree? Does it help an actual use case?

Reply:
Yes, with pushdown predicate if we have partition information on `AA` which would allow skipping all files under it. This will be more clear in my next patch (and I'll add an explicit test case).